--
title: "General PlotTracks Code"
author: "asj"
date: "11/25/2014"
output: html_document
---

##Required library
```{r}
library ("Gviz")
library ("GenomicRanges")
library ("Rsamtools")
library("RSQLite")
library("DBI")
#library("biomaRt")
#library("TxDb.Hsapiens.UCSC.hg19.knownGene")
```

##Required info

```{r}
plots_dir <- "~/asj_cpd/production/plotracks/inma_20151001/plots"

#genome <- "hg19"
genome <- "hg38"

#txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene


wanted_genes.df <- data.frame (gene = c("CCND3", "IDI2", "IGFBP1", "LAMA3", "MYH15", "PHTF1", "PIK3C2G", "RAD52", "VAMP1", "ZNF813"), ensg_id = c("ENSG00000112576", "ENSG00000148377", "ENSG00000146678", "ENSG00000053747", "ENSG00000144821", "ENSG00000116793", "ENSG00000139144", "ENSG00000002016", "ENSG00000139190", "ENSG00000198346"), is_case = c("0", "1", "1", "0", "0", "1", "1", "0", "1", "0")) 

bamFiles <- c("/storage/ethernus_backup/exome_400_workdir/tmp/MET_400.merge.coincident_variants.subset_chr.sorted.bam")

group <- 1 ##cases
#group <- 0 ##controls

query <- wanted_genes.df[wanted_genes.df$is_case == group,]$ensg_id

```

##Getting gene info from sqlite
```{r}

db_path <- "/cpd_home/pmg/databases/ensembl_v75.db"

con <- dbConnect(RSQLite::SQLite(), dbname = db_path)

##List all dataTables
#alltables <- dbListTables(con)

info_query_genes.db <- NULL

for (i in query){
  # test: i <- "ENSG00000112576"
  if (is.data.frame(info_query_genes.db) == FALSE){
    info_query_genes.db <- dbGetQuery(con, paste0('select * from ens_exon where gene_id == \"', i, "\";"))
    }else{
      new_query <- dbGetQuery(con, paste0('select * from ens_exon where gene_id == \"', i, "\";"))
      info_query_genes.db <- rbind(info_query_genes.db, new_query)
    }
  }

```

##PlotTracks function
```{r}

final_plotrack <- function(query){
  
  query_result <- info_query_genes.db
  exon_width <- query_result$end - query_result$start
  query_result <- cbind(query_result, exon_width)[,c(1:8,12)]
  names(query_result) <- c("gene", "symbol", "transcript", "exon", "chromosome", "start", "end",  "strand", "exon_width")
  query_result$strand <- ifelse(query_result$strand==1,  "+", "-")
  
  
  for (gene in query) {
    #test: gene <- "ENSG00000139190"
    
    query_result_subset <- query_result[query_result$gene == gene,]
    chr <- unique(query_result_subset[query_result_subset$gene == gene, "chromosome"])
    start_gene <- min(query_result_subset[query_result_subset$gene == gene, "start"])
    end_gene <- max(query_result_subset[query_result_subset$gene == gene, "end"])
    
    start <- start_gene - 1500
    end <- end_gene + 1500  
    
    iTrack <- IdeogramTrack(genome = genome, chromosome = chr)
    
    #gTrack <- GenomeAxisTrack(add53 = T, add35 = T)
    
    
    gmtrack <- GeneRegionTrack (  query_result_subset  
                                  , genome = genome
                                  , chromosome = chr
                                  , from = start
                                  , to = end
                                  , transcriptAnnotation = "transcript"
                                  , showId = TRUE 
                                  , name = "Transcripts"
                                  )
      
    if (length(bamFiles) == 1){
      bamTracks <- DataTrack(          range = bamFiles
                                       , genome = genome
                                       , type = c("g", "l") #c("l","g")
                                       , name = "Coverage"
                                       , chromosome = chr
                                       , from = start
                                       , to = end
                                       #, ylim = c(0,150)
                                       #     , background.panel = final_background
                                       )
      
#       alignmentTrack <- AlignmentsTrack( range = bamFiles
#                                        , from = start
#                                        , to = end
#                                        , name = "AlignmentsTrack"
#                                        , isPaired = TRUE
#                                        , ucscChromosomeNames=FALSE
#                                    #    , chromosome = "chr5"                                    
#                                        )
      
      png(file=file.path(plots_dir, paste0(gene, "_plotTrack.png")), width=500, heigh=100, units="mm", res=300)
       
      plotTracks(list( iTrack, 
                       # gTrack,
                       gmtrack,
                       bamTracks
                      # , alignmentTrack
                       )
                 , main = gene
                 )
      
      }else{
        
        bamTracks <-list()
        
        for (bam in bamFiles){
          
          bamTracks[[bam]] <- DataTrack( range = bam
                                         , genome = genome
                                         , type = c("l","g")
                                         , name = bam
                                         , chromosome = chr
                                         , from = start
                                         , to = end
                                         , ylim = c(0,150)
                                         #     , background.panel = final_background
                                         )
          }
        
        png(file=file.path(plots_dir, paste0(gene, "_plotTrack.png")), width=500, heigh=100, units="mm", res=300)
        
        plotTracks(c(list(
          gTrack,
          gmtrack),
          bamTracks
          )
          , from = start
          , end = end
          , main = gene
          )
        }
    
    
    dev.off()
    }
  }

final_plots <- final_plotrack (query)

```


